docs/development/best-practices.md

```diff
@@ -0,0 +1,139 @@
+Otherwise, inline metadata into the log message if logging at `INFO` or below, or if the metadata object is complex.
+
+`WARN` and above messages are often used in metrics or error catching services, and should have a consistent `msg` component so that they will be automatically grouped/associated together.
+Metadata which is separate from its message is harder for human readability, so try to combine it in the message unless it's too complex to do so.
+
+Good:
+
+```js
+logger.debug({ config }, 'Full config');
+logger.warn({ presetName }, 'Failed to look up preset');
+```
+
+Avoid:
+
+```js
+logger.debug({ branchName }, 'Generated branchName');
+logger.warn(`Failed to look up preset ${presetName}`);
+```
+
+### Array constructor
+
+Avoid the `Array()` constructor, with or without `new`, in your TypeScript code.
+It has confusing and contradictory usage.
+So you should avoid:
+
+```js
+const a = new Array(2); // [undefined, undefined]
+const b = new Array(2, 3); // [2, 3];
+```
+
+Instead, always use bracket notation to initialize arrays, or `from` to initialize an Array with a certain size i.e.
+
+```js
+// [0, 0, 0, 0, 0]
+Array.from < number > { length: 5 }.fill(0);
+```
+
+[Source](https://google.github.io/styleguide/tsguide.html#array-constructor)
+
+### Iterating objects & containers
+
+Use `for ( ... of ...)` loops instead of `[Array|Set|Map].prototype.forEach` and `for ( ... in ...)`.
+
+- Using `for ( ... in ...)` for objects is error-prone. It will include enumerable properties from the prototype chain
+- Using `for ( ... in ...)` to iterate over arrays, will counterintuitively return the array's indices
+- Avoid `[Array|Set|Map].prototype.forEach`. It makes code harder to debug and defeats some useful compiler checks like reachability
+
+Only use `Array.prototype.map()` when the return value is used, otherwise use `for ( ... of ...)`.
+
+[Source](https://google.github.io/styleguide/tsguide.html#iterating-objects)
+
+### Exports
+
+Use named exports in all code.
+Avoid default `exports`.
+This way all `imports` follow the same pattern.
+
+[Source, reasoning and examples.](https://google.github.io/styleguide/tsguide.html#exports)
+
+### Imports
+
+Use [ES6 module](https://exploringjs.com/es6/ch_modules.html#sec_basics-of-es6-modules) syntax, i.e.
+
+```js
+import { square, diag } from 'lib';
+
+// You may also use:
+
+import * as lib from 'lib';
+```
+
+And avoid `require`:
+
+```js
+import x = require('...');
+```
+
+### HTTP & RESTful API request handling
+
+Prefer using [`got`](https://www.npmjs.com/package/got) to simplify HTTP request handling.
+
+### Dates and times
+
+Use [`Luxon`](https://www.npmjs.com/package/luxon) to handle dates and times.
+Use `UTC` to be time zone independent.
+
+[Example](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/distro.ts#L133-L134)
+
+### Unit testing
+
+- Use `it.each` rather than `test.each`
```

Comment:
```suggestion
- Use `it.each` rather than `test.each`
  - Prefer  [Tagged Template Literal](https://jestjs.io/docs/api#2-testeachtablename-fn-timeout) style for `it.each`, prettier will help with format
```